File: src/net/dinkla/raytracer/math/Normal.java
///*
//package net.dinkla.raytracer.math;
//
//import static java.lang.Math.sqrt;
//
//*/
///**
// * Created by IntelliJ IDEA.
// * User: jorndinkla
// * Date: 10.04.2010
// * Time: 15:17:52
// * To change this template use File | Settings | File Templates.
// */
//
//public class Normal extends Element3D {
//
//    public static final Normal RIGHT = new Normal(1, 0, 0);
//    public static final Normal LEFT = new Normal(-1, 0, 0);
//    public static final Normal UP = new Normal(0, 1, 0);
//    public static final Normal DOWN = new Normal(0, -1, 0);
//    public static final Normal FRONT = new Normal(0, 0, 1);
//    public static final Normal BACK = new Normal(0, 0, -1);
//    public static final Normal ZERO = new Normal(0, 0, 0);
//
//    public Normal(float x, float y, float z) {
//        super(x, y, z);
//    }
//
//    public Normal(final Vector3DF v) {
//        super(v.normalize());
//    }
//
//    public Normal(final Point3DF p0, final Point3DF p1, final Point3DF p2) {
//        super(p1.minus(p0).cross(p2.minus(p0)).normalize());
//    }
//
//    public Normal normalize() {
//        float len = length();
//        return new Normal(x / len, y / len, z / len);
//    }
//
//    public Vector3DF mult(final float s) {
//        return new Vector3DF(s * x, s * y, s * z);
//    }
//
//    public Vector3DF plus(final Normal normal) {
//        return new Vector3DF(x + normal.x, y + normal.y, z + normal.z);
//    }
//
//    public float dot(final Vector3DF v)  {
//        return x * v.x + y * v.y + z * v.z;
//    }
//
//    public Normal negate() {
//        return new Normal(-x, -y, -z);
//    }
//
//    @Override
//    public Object x() {
//        return x;
//    }
//
//    @Override
//    public Object z() {
//        return z;
//    }
//
//    @Override
//    public Object length() {
//        return sqrt(sqrLength());
//    }
//
//    @Override
//    public Object y() {
//        return y;
//    }
//}

File: src/net/dinkla/raytracer/math/Normal.kt
package net.dinkla.raytracer.math

class Normal : Element3D {

    constructor(x: Double, y: Double, z: Double) : super(x, y, z) {}

    constructor(v: Vector3D) : super(v.normalize()) {}

    constructor(p0: Point3D, p1: Point3D, p2: Point3D) : super(p1.minus(p0).cross(p2.minus(p0)).normalize()) {}

    fun normalize(): Normal {
        val len = length()
        return Normal(x / len, y / len, z / len)
    }

    fun mult(s: Double): Vector3D {
        return Vector3D(s * x, s * y, s * z)
    }

    operator fun plus(normal: Normal): Vector3D {
        return Vector3D(x + normal.x, y + normal.y, z + normal.z)
    }

    fun dot(v: Vector3D): Double {
        return x * v.x + y * v.y + z * v.z
    }

    fun negate(): Normal {
        return Normal(-x, -y, -z)
    }

    companion object {
        val RIGHT = Normal(1.0, 0.0, 0.0)
        val ONE = Normal(1.0, 1.0, 1.0)
        val LEFT = Normal(-1.0, 0.0, 0.0)
        val UP = Normal(0.0, 1.0, 0.0)
        val DOWN = Normal(0.0, -1.0, 0.0)
        val FRONT = Normal(0.0, 0.0, 1.0)
        val BACK = Normal(0.0, 0.0, -1.0)
        val ZERO = Normal(0.0, 0.0, 0.0)
    }
}
